### PR TITLE
UniswapV2Factory full test

### LINKED ISSUE #467: feat: Fix UniswapV2Factory tests
# Feature Request

**Describe the Feature Request**

When deploying the `UniswapV2Factory` we face a couple of bugs that have been addressed in dedicated PR. We are now in a position to pass these tests.

**Describe Preferred Solution**

The reference file for this test is https://github.com/Uniswap/v2-core/blob/master/test/UniswapV2Factory.spec.ts
The corresponding file in kakarot is https://github.com/sayajin-labs/kakarot/blob/main/tests/integration/solidity_contracts/UniswapV2/test_uniswap_v2_factory.py

- set the `TEST_ADDRESSES` to what they are in the original file 

```
const TEST_ADDRESSES: [string, string] = [
  '0x1000000000000000000000000000000000000000',
  '0x2000000000000000000000000000000000000000'
]
```

Unskip `TestCreatePair` and make it pass!

You'll also need to 

- move the `deployer` fixture in `tests/fixtures/2_kakarot.py`
- update `get_solidity_contract` fixture (see `tests/integration/conftest.py`):
    - add the `deployer` fixture in the args
    - make the `starknet_contract_address=None`
    - make the `tx=None`
    - if `starknet_contract_address is None` get the address from the deployer

### PR DESCRIPTION
**Pull request by ClementWalter**
Time spent on this PR: 0.5

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Token addresses are random instead the same one as in the original test.
Test using `createPair` is skipped

Resolves #467 #452 

## What is the new behavior?

All tests pass as in the original file

### CODE REVIEW

File: src/kakarot/instructions/environmental_information.cairo [hunk 1]
```diff
@@ -163,7 +163,9 @@ namespace EnvironmentalInformation {
         alloc_locals;
         // Get caller address.
         let (current_address) = get_caller_address();
-        let (evm_address) = IAccount.get_eth_address(current_address);
+        let (current_evm_address) = IAccount.get_eth_address(current_address);
+        let is_root = ExecutionContext.is_root(ctx);
+        let evm_address = (1 - is_root) * ctx.calling_context.evm_contract_address + is_root * current_evm_address;
```
Comment (Eikix) [inline src/kakarot/instructions/environmental_information.cairo:168]: nice catch!

**Review by Eikix — APPROVED** [review]
lgtm

**Review by Eikix — CHANGES_REQUESTED** [review]
Nvm I didn't notice tests were failing

File: src/kakarot/instructions/environmental_information.cairo (hunk 1 above)
Comment (ClementWalter) [reply src/kakarot/instructions/environmental_information.cairo:168]: tribute to @seshanthS actually

**Comment by ClementWalter** [comment]
> Nvm I didn't notice tests were failing

It took me quite a lot of time to understand what was the issue. Actually the `ExecutionContext.is_root` and `ExecutionContext.is_leaf` (note: I wrote them myself two months ago) were not checking that a given _true_ context was root or leaf ; it was working only on the dummy `empty_ctx`.

These methods are not tested (!) and actually `is_root` was used only once with 

```
is_parent_root = ExecutionContext.is_root(ctx.calling_context);
```

which was quite misleading imao against past me. Consequently I've updated these helpers to be usable on the actual current execution context. For now, I've just considered that checking against an empty `evm_contract_address` was enough, wdyt?

**Comment by Eikix** [comment]
> 

Are we still maintaining kakarot.execute?

It looked to me as though we were slowly deprecating kakarot.execute in favor of kakarot.execute_at_address

**Pull request merged**